File: core/safety.py
"""연결 파일 감지: 시스템 경로, 바로가기 대상, 레지스트리 등록 경로, 잠김, 프로젝트 폴더, 클라우드 자리표시자.

risk 등급
  block : 기본적으로 절대 옮기지 않음 (시스템 폴더, 프로그램 폴더, 레지스트리 등록, 잠김)
  warn  : 옮길 수는 있으나 확인 필요 (바로가기 대상, 프로젝트 폴더 내부, 클라우드 자리표시자, 실행파일)
  safe  : 자유롭게 정리 가능
"""
from __future__ import annotations

import os
import struct
import sys
from functools import lru_cache

from .scanner import FileInfo

IS_WIN = sys.platform == "win32"

EXEC_EXT = {"exe", "dll", "sys", "msi", "com", "bat", "cmd", "ps1", "vbs", "scr", "ocx", "drv", "cpl"}
PROJECT_MARKERS = {
    ".git", ".svn", ".hg", "package.json", "pyproject.toml", "requirements.txt", "Cargo.toml",
    "go.mod", "pom.xml", "build.gradle", "CMakeLists.txt", "Makefile", ".sln", ".csproj", ".vcxproj",
    "Gemfile", "composer.json", ".vs", ".idea", "node_modules", "venv", ".venv",
}
CLOUD_HINTS = ("onedrive", "dropbox", "google drive", "googledrive", "icloud", "box sync")


def _env(name: str) -> str:
    return os.environ.get(name, "")


@lru_cache(maxsize=1)
def system_roots() -> list[str]:
    roots = [
        _env("SystemRoot"), _env("ProgramFiles"), _env("ProgramFiles(x86)"), _env("ProgramData"),
        _env("LOCALAPPDATA"), _env("APPDATA"), _env("ProgramW6432"),
    ]
    drive = os.path.splitdrive(_env("SystemRoot") or "C:\\")[0]
    roots += [drive + "\\$Recycle.Bin", drive + "\\System Volume Information", drive + "\\Recovery",
              drive + "\\PerfLogs", drive + "\\Windows"]
    return [os.path.normcase(os.path.abspath(r)) for r in roots if r]


def _under(path: str, base: str) -> bool:
    p = os.path.normcase(os.path.abspath(path))
    return p == base or p.startswith(base.rstrip("\\") + "\\")


# ---------------------------------------------------------------- .lnk 파싱 (순수 파이썬)
def parse_lnk_target(path: str) -> str | None:
    """Windows 바로가기(.lnk)의 대상 경로를 읽는다. 실패하면 None."""
    try:
        with open(path, "rb") as f:
            data = f.read(64 * 1024)
        if len(data) < 0x4C or data[:4] != b"L\x00\x00\x00":
            return None
        flags = struct.unpack_from("<I", data, 0x14)[0]
        off = 0x4C
        if flags & 0x1:  # HasLinkTargetIDList
            idlist_size = struct.unpack_from("<H", data, off)[0]
            off += 2 + idlist_size
        if flags & 0x2:  # HasLinkInfo
            li_size, li_hdr, li_flags = struct.unpack_from("<III", data, off)
            local_off = struct.unpack_from("<I", data, off + 16)[0]
            if li_flags & 0x1 and local_off:
                start = off + local_off
                end = data.index(b"\x00", start)
                raw = data[start:end]
                try:
                    return raw.decode("mbcs")
                except Exception:
                    return raw.decode("utf-8", "ignore")
            # 네트워크 경로 등은 생략
        return None
    except Exception:
        return None


def shortcut_dirs() -> list[str]:
    home = os.path.expanduser("~")
    cands = [
        os.path.join(home, "Desktop"), os.path.join(home, "OneDrive", "Desktop"),
        os.path.join(_env("APPDATA"), "Microsoft", "Windows", "Start Menu"),
        os.path.join(_env("ProgramData"), "Microsoft", "Windows", "Start Menu"),
        os.path.join(_env("APPDATA"), "Microsoft", "Internet Explorer", "Quick Launch"),
        os.path.join(_env("PUBLIC"), "Desktop"),
    ]
    return [c for c in cands if c and os.path.isdir(c)]


def collect_shortcut_targets(extra_dirs: list[str] | None = None) -> dict[str, str]:
    """{대상경로(normcase): 바로가기 경로}"""
    out: dict[str, str] = {}
    for d in shortcut_dirs() + (extra_dirs or []):
        for cur, _dirs, names in os.walk(d):
            for n in names:
                if n.lower().endswith(".lnk"):
                    lp = os.path.join(cur, n)
                    t = parse_lnk_target(lp)
                    if t:
                        out[os.path.normcase(os.path.abspath(t))] = lp
    return out


# ---------------------------------------------------------------- 레지스트리
def collect_registry_paths() -> set[str]:
    """레지스트리에 등록된 프로그램 경로(normcase)와 그 상위 폴더."""
    paths: set[str] = set()
    if not IS_WIN:
        return paths
    import winreg

    def add(val: str):
        if not isinstance(val, str) or ":" not in val:
            return
        v = val.strip().strip('"')
        # "C:\path\app.exe" --arg 형태에서 경로만
        if v.lower().endswith(".exe") or os.path.exists(v):
            p = os.path.normcase(os.path.abspath(v))
        else:
            idx = v.lower().find(".exe")
            if idx == -1:
                return
            p = os.path.normcase(os.path.abspath(v[: idx + 4].strip('"')))
        paths.add(p)
        paths.add(os.path.dirname(p))

    keys = [
        (winreg.HKEY_LOCAL_MACHINE, r"SOFTWARE\Microsoft\Windows\CurrentVersion\App Paths"),
        (winreg.HKEY_CURRENT_USER, r"SOFTWARE\Microsoft\Windows\CurrentVersion\App Paths"),
        (winreg.HKEY_LOCAL_MACHINE, r"SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall"),
        (winreg.HKEY_LOCAL_MACHINE, r"SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall"),
        (winreg.HKEY_CURRENT_USER, r"SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall"),
        (winreg.HKEY_LOCAL_MACHINE, r"SOFTWARE\Microsoft\Windows\CurrentVersion\Run"),
        (winreg.HKEY_CURRENT_USER, r"SOFTWARE\Microsoft\Windows\CurrentVersion\Run"),
    ]
    for hive, sub in keys:
        try:
            k = winreg.OpenKey(hive, sub)
        except OSError:
            continue
        with k:
            # 값들
            i = 0
            while True:
                try:
                    _n, v, _t = winreg.EnumValue(k, i)
                    add(v)
                    i += 1
                except OSError:
                    break
            # 하위 키들
            i = 0
            while True:
                try:
                    name = winreg.EnumKey(k, i)
                    i += 1
                except OSError:
                    break
                try:
                    with winreg.OpenKey(k, name) as sk:
                        for vn in ("", "Path", "InstallLocation", "DisplayIcon", "UninstallString"):
                            try:
                                v, _t = winreg.QueryValueEx(sk, vn)
                                add(v)
                            except OSError:
                                pass
                except OSError:
                    pass
    return paths


def is_locked(path: str) -> bool:
    """다른 프로세스가 열어 두었는지 검사 (독점 열기 시도)."""
    if not IS_WIN:
        return False
    try:
        import msvcrt
        with open(path, "rb") as f:
            try:
                msvcrt.locking(f.fileno(), msvcrt.LK_NBLCK, 1)
                msvcrt.locking(f.fileno(), msvcrt.LK_UNLCK, 1)
            except OSError:
                return True
    except PermissionError:
        return True
    except OSError:
        return False
    return False


# ---------------------------------------------------------------- 분석기
class SafetyAnalyzer:
    def __init__(self, root: str, check_locks: bool = True, lock_check_limit: int = 5000):
        self.root = os.path.abspath(root)
        self.sys_roots = system_roots()
        self.shortcuts = collect_shortcut_targets([self.root])
        self.shortcut_dirs = {os.path.dirname(p) for p in self.shortcuts}
        self.reg_paths = collect_registry_paths()
        self.check_locks = check_locks
        self.lock_check_limit = lock_check_limit
        self._project_cache: dict[str, str | None] = {}
        self._program_dir_cache: dict[str, bool] = {}

    def project_root_of(self, directory: str) -> str | None:
        """directory 또는 그 상위(루트까지)에 프로젝트 표식이 있으면 그 폴더."""
        d = directory
        chain = []
        while True:
            if d in self._project_cache:
                res = self._project_cache[d]
                break
            chain.append(d)
            try:
                names = set(os.listdir(d))
            except OSError:
                names = set()
            if names & PROJECT_MARKERS or any(n.endswith((".sln", ".csproj")) for n in names):
                res = d
                break
            parent = os.path.dirname(d)
            if parent == d or os.path.normcase(d) == os.path.normcase(self.root):
                res = None
                break
            d = parent
        for c in chain:
            self._project_cache[c] = res
        return res

    def is_program_dir(self, directory: str) -> bool:
        """폴더에 exe와 dll이 함께 있으면 설치된 프로그램 폴더로 본다."""
        if directory in self._program_dir_cache:
            return self._program_dir_cache[directory]
        try:
            exts = {n.rsplit(".", 1)[-1].lower() for n in os.listdir(directory) if "." in n}
        except OSError:
            exts = set()
        res = "exe" in exts and ("dll" in exts or "manifest" in exts)
        self._program_dir_cache[directory] = res
        return res

    def analyze(self, files: list[FileInfo], progress=None) -> dict:
        stats = {"safe": 0, "warn": 0, "block": 0}
        lock_checked = 0
        for i, f in enumerate(files):
            reasons: list[str] = []
            level = "safe"
            p = os.path.normcase(f.path)
            d = os.path.dirname(f.path)
            group = d

            if any(_under(f.path, r) for r in self.sys_roots):
                level = "block"; reasons.append("reason.sysdir")
            if p in self.reg_paths or os.path.normcase(d) in self.reg_paths:
                level = "block"; reasons.append("reason.registry")
            if f.system:
                level = "block"; reasons.append("reason.sysattr")
            if self.is_program_dir(d):
                level = "block"; reasons.append("reason.programdir"); group = d

            if level != "block":
                if p in self.shortcuts:
                    level = "warn"; reasons.append("reason.shortcut|" + os.path.basename(self.shortcuts[p]))
                if f.ext in EXEC_EXT:
                    level = "warn"; reasons.append("reason.exec")
                if f.ext == "lnk":
                    level = "warn"; reasons.append("reason.lnk")
                pr = self.project_root_of(d)
                if pr:
                    level = "warn"; reasons.append("reason.project|" + os.path.basename(pr)); group = pr
                if f.cloud_placeholder or any(h in p for h in CLOUD_HINTS):
                    level = "warn"; reasons.append("reason.cloud")
                if f.hidden:
                    level = "warn"; reasons.append("reason.hidden")
                if self.check_locks and lock_checked < self.lock_check_limit and f.ext not in ("lnk",) and not f.cloud_placeholder:
                    lock_checked += 1
                    if is_locked(f.path):
                        level = "block"; reasons.append("reason.locked")

            f.risk = level
            f.reasons = reasons
            f.group = group
            stats[level] += 1
            if progress and i % 2000 == 0:
                progress(i)
        return {
            "stats": stats,
            "shortcut_count": len(self.shortcuts),
            "registry_count": len(self.reg_paths),
        }

File: FoldeClean/Core/Safety.cs
using System.IO;
using System.Text;
using Microsoft.Win32;

namespace FoldeClean;

/// <summary>
/// 연결 파일 감지.
///   block : 절대 옮기지 않음 (시스템 폴더, 프로그램 폴더, 레지스트리 등록, 잠김)
///   warn  : 확인 필요 (바로가기 대상, 프로젝트 내부, 클라우드 자리표시자, 실행파일, 숨김)
///   safe  : 자유롭게 정리 가능
/// </summary>
public sealed class SafetyAnalyzer
{
    static readonly HashSet<string> ExecExt = new() { "exe", "dll", "sys", "msi", "com", "bat", "cmd", "ps1", "vbs", "scr", "ocx", "drv", "cpl" };
    static readonly HashSet<string> ProjectMarkers = new(StringComparer.OrdinalIgnoreCase) {
        ".git", ".svn", ".hg", "package.json", "pyproject.toml", "requirements.txt", "Cargo.toml", "go.mod", "pom.xml",
        "build.gradle", "CMakeLists.txt", "Makefile", "Gemfile", "composer.json", ".vs", ".idea", "node_modules", "venv", ".venv" };
    static readonly string[] CloudHints = { "onedrive", "dropbox", "google drive", "googledrive", "icloud", "box sync" };

    readonly string _root;
    readonly List<string> _sysRoots;
    readonly Dictionary<string, string> _shortcuts;      // 대상경로(lower) → lnk 경로
    readonly HashSet<string> _regPaths;
    readonly bool _checkLocks;
    readonly int _lockLimit;
    readonly Dictionary<string, string?> _projectCache = new(StringComparer.OrdinalIgnoreCase);
    readonly Dictionary<string, bool> _programDirCache = new(StringComparer.OrdinalIgnoreCase);

    public int ShortcutCount => _shortcuts.Count;
    public int RegistryCount => _regPaths.Count;

    public SafetyAnalyzer(string root, bool checkLocks = true, int lockCheckLimit = 5000, int scanDepth = 50)
    {
        _root = Path.GetFullPath(root);
        _sysRoots = SystemRoots();
        _shortcuts = CollectShortcutTargets(new[] { _root }, scanDepth > 0);
        _regPaths = CollectRegistryPaths();
        _checkLocks = checkLocks;
        _lockLimit = lockCheckLimit;
    }

    // ------------------------------------------------------------ 시스템 경로
    static string Env(string n) => Environment.GetEnvironmentVariable(n) ?? "";

    static List<string> SystemRoots()
    {
        var roots = new List<string> { Env("SystemRoot"), Env("ProgramFiles"), Env("ProgramFiles(x86)"), Env("ProgramData"),
            Env("LOCALAPPDATA"), Env("APPDATA"), Env("ProgramW6432") };
        var drive = Path.GetPathRoot(Env("SystemRoot").Length > 0 ? Env("SystemRoot") : @"C:\") ?? @"C:\";
        roots.AddRange(new[] { "$Recycle.Bin", "System Volume Information", "Recovery", "PerfLogs", "Windows" }.Select(n => Path.Combine(drive, n)));
        return roots.Where(r => r.Length > 0).Select(r => Norm(Path.GetFullPath(r))).ToList();
    }

    static string Norm(string p) => p.TrimEnd('\\').ToLowerInvariant();
    static bool Under(string path, string baseNorm) { var p = Norm(path); return p == baseNorm || p.StartsWith(baseNorm + "\\"); }

    // ------------------------------------------------------------ .lnk 파싱
    public static string? ParseLnkTarget(string path)
    {
        try
        {
            byte[] data;
            using (var f = File.OpenRead(path)) { data = new byte[Math.Min(f.Length, 64 * 1024)]; f.ReadExactly(data); }
            if (data.Length < 0x4C || data[0] != 'L' || data[1] != 0 || data[2] != 0 || data[3] != 0) return null;
            uint flags = BitConverter.ToUInt32(data, 0x14);
            int off = 0x4C;
            if ((flags & 0x1) != 0) { int idSize = BitConverter.ToUInt16(data, off); off += 2 + idSize; }
            if ((flags & 0x2) != 0)
            {
                uint liFlags = BitConverter.ToUInt32(data, off + 8);
                uint localOff = BitConverter.ToUInt32(data, off + 16);
                if ((liFlags & 0x1) != 0 && localOff != 0)
                {
                    int start = off + (int)localOff;
                    int end = Array.IndexOf(data, (byte)0, start);
                    if (end < 0) return null;
                    return Encoding.Default.GetString(data, start, end - start);
                }
            }
            return null;
        }
        catch { return null; }
    }

    static IEnumerable<string> ShortcutDirs()
    {
        var home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        var c = new[] {
            Environment.GetFolderPath(Environment.SpecialFolder.Desktop), Path.Combine(home, "OneDrive", "Desktop"),
            Environment.GetFolderPath(Environment.SpecialFolder.StartMenu), Environment.GetFolderPath(Environment.SpecialFolder.CommonStartMenu),
            Path.Combine(Env("APPDATA"), "Microsoft", "Internet Explorer", "Quick Launch"),
            Environment.GetFolderPath(Environment.SpecialFolder.CommonDesktopDirectory) };
        return c.Where(d => d.Length > 0 && Directory.Exists(d));
    }

    /// <summary>
    /// 바로가기가 가리키는 파일을 모은다. 바탕화면·시작 메뉴는 늘 하위까지 보고,
    /// 정리 대상 폴더(root)는 검사 범위와 같은 깊이만 본다.
    /// 예전에는 root 도 무조건 하위 전체를 훑어, "이 폴더의 파일만" 검사인데도 오래 걸렸다.
    /// </summary>
    static Dictionary<string, string> CollectShortcutTargets(IEnumerable<string> extra, bool extraRecursive)
    {
        var map = new Dictionary<string, string>();
        var deep = new EnumerationOptions { IgnoreInaccessible = true, RecurseSubdirectories = true, AttributesToSkip = FileAttributes.ReparsePoint };
        var flat = new EnumerationOptions { IgnoreInaccessible = true, RecurseSubdirectories = false, AttributesToSkip = FileAttributes.ReparsePoint };

        void Collect(string dir, EnumerationOptions opts)
        {
            IEnumerable<string> files;
            try { files = Directory.EnumerateFiles(dir, "*.lnk", opts); } catch { return; }
            foreach (var lp in files)
            {
                var t = ParseLnkTarget(lp);
                if (t != null) { try { map[Norm(Path.GetFullPath(t))] = lp; } catch { } }
            }
        }

        foreach (var d in ShortcutDirs()) Collect(d, deep);
        foreach (var d in extra) Collect(d, extraRecursive ? deep : flat);
        return map;
    }

    // ------------------------------------------------------------ 레지스트리
    static HashSet<string> CollectRegistryPaths()
    {
        var paths = new HashSet<string>();
        void Add(object? val)
        {
            if (val is not string s || !s.Contains(':')) return;
            var v = s.Trim().Trim('"');
            string p;
            if (v.EndsWith(".exe", StringComparison.OrdinalIgnoreCase) || File.Exists(v) || Directory.Exists(v)) p = v;
            else { int i = v.IndexOf(".exe", StringComparison.OrdinalIgnoreCase); if (i < 0) return; p = v[..(i + 4)].Trim('"'); }
            try { p = Norm(Path.GetFullPath(p)); } catch { return; }
            paths.Add(p);
            var dir = Path.GetDirectoryName(p); if (dir != null) paths.Add(Norm(dir));
        }
        var keys = new (RegistryKey hive, string sub)[] {
            (Registry.LocalMachine, @"SOFTWARE\Microsoft\Windows\CurrentVersion\App Paths"),
            (Registry.CurrentUser, @"SOFTWARE\Microsoft\Windows\CurrentVersion\App Paths"),
            (Registry.LocalMachine, @"SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall"),
            (Registry.LocalMachine, @"SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall"),
            (Registry.CurrentUser, @"SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall"),
            (Registry.LocalMachine, @"SOFTWARE\Microsoft\Windows\CurrentVersion\Run"),
            (Registry.CurrentUser, @"SOFTWARE\Microsoft\Windows\CurrentVersion\Run") };
        foreach (var (hive, sub) in keys)
        {
            RegistryKey? k = null;
            try { k = hive.OpenSubKey(sub); } catch { }
            if (k == null) continue;
            using (k)
            {
                foreach (var vn in k.GetValueNames()) { try { Add(k.GetValue(vn)); } catch { } }
                foreach (var name in k.GetSubKeyNames())
                {
                    try
                    {
                        using var sk = k.OpenSubKey(name);
                        if (sk == null) continue;
                        foreach (var vn in new[] { "", "Path", "InstallLocation", "DisplayIcon", "UninstallString" })
                            try { Add(sk.GetValue(vn)); } catch { }
                    }
                    catch { }
                }
            }
        }
        return paths;
    }

    // ------------------------------------------------------------ 잠김
    /// <summary>읽기 전용 + 공유 금지로 열어 본다. 다른 프로세스가 열어 두었으면 실패한다. (쓰기 열기는 백신 검사를 유발해 느리다)</summary>
    static bool IsLocked(string path)
    {
        try { using var f = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.None, 1, FileOptions.None); return false; }
        catch (IOException) { return true; }
        catch (UnauthorizedAccessException) { return true; }
        catch { return false; }
    }

    const int AttrOffline = 0x1000;
    public int LockChecked { get; private set; }
    public bool LockCheckTimedOut { get; private set; }

    // ------------------------------------------------------------ 프로젝트 / 프로그램 폴더
    string? ProjectRootOf(string dir)
    {
        var chain = new List<string>();
        string? res = null;
        var d = dir;
        while (true)
        {
            if (_projectCache.TryGetValue(d, out res)) break;
            chain.Add(d);
            bool marker = false;
            try
            {
                foreach (var e in Directory.EnumerateFileSystemEntries(d))
                {
                    var n = Path.GetFileName(e);
                    if (ProjectMarkers.Contains(n) || n.EndsWith(".sln", StringComparison.OrdinalIgnoreCase) || n.EndsWith(".csproj", StringComparison.OrdinalIgnoreCase)) { marker = true; break; }
                }
            }
            catch { }
            if (marker) { res = d; break; }
            var parent = Path.GetDirectoryName(d);
            if (parent == null || string.Equals(d, _root, StringComparison.OrdinalIgnoreCase)) { res = null; break; }
            d = parent;
        }
        foreach (var c in chain) _projectCache[c] = res;
        return res;
    }

    bool IsProgramDir(string dir)
    {
        if (_programDirCache.TryGetValue(dir, out var r)) return r;
        bool exe = false, dll = false;
        try
        {
            foreach (var e in Directory.EnumerateFiles(dir))
            {
                var x = Path.GetExtension(e).ToLowerInvariant();
                if (x == ".exe") exe = true; else if (x == ".dll" || x == ".manifest") dll = true;
                if (exe && dll) break;
            }
        }
        catch { }
        return _programDirCache[dir] = exe && dll;
    }

    // ------------------------------------------------------------ 분석
    public Dictionary<string, object?> Analyze(List<FileRec> files, Action<int>? progress = null, Action<int, int>? lockProgress = null, int lockTimeBudgetSec = 20)
    {
        int safe = 0, warn = 0, block = 0;
        for (int i = 0; i < files.Count; i++)
        {
            var f = files[i];
            var reasons = new List<string>();
            var level = "safe";
            var p = Norm(f.Path);
            var d = Path.GetDirectoryName(f.Path) ?? "";

            string group = d;   // 기본 묶음: 상위 폴더
            if (_sysRoots.Any(r => Under(f.Path, r))) { level = "block"; reasons.Add("reason.sysdir"); }
            if (_regPaths.Contains(p) || _regPaths.Contains(Norm(d))) { level = "block"; reasons.Add("reason.registry"); }
            if (f.System) { level = "block"; reasons.Add("reason.sysattr"); }
            if (IsProgramDir(d)) { level = "block"; reasons.Add("reason.programdir"); group = d; }

            if (level != "block")
            {
                if (_shortcuts.TryGetValue(p, out var lnk)) { level = "warn"; reasons.Add("reason.shortcut|" + Path.GetFileName(lnk)); }
                if (ExecExt.Contains(f.Ext)) { level = "warn"; reasons.Add("reason.exec"); }
                if (f.Ext == "lnk") { level = "warn"; reasons.Add("reason.lnk"); }
                var pr = ProjectRootOf(d);
                if (pr != null) { level = "warn"; reasons.Add("reason.project|" + Path.GetFileName(pr)); group = pr; }
                if (f.CloudPlaceholder || CloudHints.Any(h => p.Contains(h))) { level = "warn"; reasons.Add("reason.cloud"); }
                if (f.Hidden) { level = "warn"; reasons.Add("reason.hidden"); }
            }
            f.Risk = level; f.Reasons = reasons; f.Group = group;
            if (progress != null && i % 200 == 0) progress(i);
        }

        // 잠김 검사: 병렬 + 시간 제한. 클라우드 자리표시자/오프라인 파일은 열면 내려받기가 시작되므로 건너뛴다.
        if (_checkLocks)
        {
            var targets = files.Where(f => f.Risk != "block" && f.Ext != "lnk" && !f.CloudPlaceholder && (f.Attrs & AttrOffline) == 0)
                               .Take(_lockLimit).ToList();
            var sw = System.Diagnostics.Stopwatch.StartNew();
            int done = 0; bool timedOut = false;
            var po = new ParallelOptions { MaxDegreeOfParallelism = 8 };
            Parallel.ForEach(targets, po, (f, state) =>
            {
                if (sw.Elapsed.TotalSeconds > lockTimeBudgetSec) { timedOut = true; state.Stop(); return; }
                if (IsLocked(f.Path)) lock (f) { f.Risk = "block"; f.Reasons.Add("reason.locked"); }
                int n = Interlocked.Increment(ref done);
                if (lockProgress != null && n % 200 == 0) lockProgress(n, targets.Count);
            });
            LockChecked = done; LockCheckTimedOut = timedOut;
        }

        foreach (var f in files) { if (f.Risk == "safe") safe++; else if (f.Risk == "warn") warn++; else block++; }
        return new()
        {
            ["stats"] = new Dictionary<string, int> { ["safe"] = safe, ["warn"] = warn, ["block"] = block },
            ["shortcut_count"] = ShortcutCount, ["registry_count"] = RegistryCount,
            ["lock_checked"] = LockChecked, ["lock_timed_out"] = LockCheckTimedOut,
        };
    }
}
